modules/lfs/shared.go

```diff
@@ -64,6 +66,39 @@ type ObjectError struct {
 	Message string `json:"message"`
 }
 
+var (
+	// See https://github.com/git-lfs/git-lfs/blob/main/docs/api/batch.md#successful-responses
+	// LFS object error codes should match HTTP status codes where possible:
+	//   404 - The object does not exist on the server.
+	//   409 - The specified hash algorithm disagrees with the server's acceptable options.
+	//   410 - The object was removed by the owner.
+	//   422 - Validation error.
+
+	ErrObjectNotFound     = errors.New("the object does not exist on the server")
+	ErrObjectHashMismatch = errors.New("the specified hash algorithm disagrees with the server's acceptable options")
+	ErrObjectRemoved      = errors.New("the object was removed by the owner")
+	ErrObjectValidation   = errors.New("validation error")
+)
+
+func (e *ObjectError) Error() string {
+	return fmt.Sprintf("[%d] %s", e.Code, e.Message)
+}
+
+func (e *ObjectError) Unwrap() error {
+	switch e.Code {
+	case 404:
+		return ErrObjectNotFound
```

Comment:
Can't we use the existing `ErrNotExist`?
I don't want to maintain a separate `NotExist` error, that'll only create problems

Reply:
Additionally, we probably want to catch `401`/`403` as `ErrPermissionDenied`

Reply:
> Can't we use the existing `ErrNotExist`?

I prefer to alias the existing `ErrNotExist`, see 7ebf151258e14af260136722195fe0fe9059ad71

So `errors.Is(objectError, lfs.ErrObjectNotExist)` and `errors.Is(objectError, util.ErrNotExist)` are both fine.

> Additionally, we probably want to catch `401`/`403` as `ErrPermissionDenied`

I don't think so, please read the [doc](https://github.com/git-lfs/git-lfs/blob/main/docs/api/batch.md?#successful-responses). There's no `401`/`403` for "object error", if there's are some permission errors, they will be "response errors" and we cannot get the objects (or the object errors).

And we need to specifically handle the permission error; just let the migration fail like handling other errors.